docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+are idle.
+
+## How (short plan)
+
+Topics will be automatically placed into the hibernate state when
+they have been idle for more than a configurable threshold.
+
+## Impact
+
+Reduced cost of delivering free/cheap public-facing services, where
+users are likely to create topics for occasional or one-off use.
+
+# Motivation
+
+## Why are we doing this?
+
+Partitions have some runtime resource overhead, even if they aren't being
+used, and haven't been used for some time.
+
+To support public-facing Kafka services on which many users might create
+and then abandon idle topics, we could implement a "hibernate" mode
+for partitions that haven't been used for a long time.
+
+## What use cases does it support?
+
+Implementing a low-cost/no-cost public service, in which developers are likely
+to create unused topics with which to experiment.
+
+## What is the expected outcome?
+
+Systems will run smoothly with a higher number of partitions than
+they could before, as long as some of those partitions are idle.
+
+# Guide-level explanation
+
+## Introducing new named concepts.
+
+`Hibernated topic`: a topic whose partitions do not have active raft
+groups running in the cluster.  It cannot serve user I/O (aside from
+metadata requests) until it is woken up.
+
+`Waking a topic`: a control plane operation to change the status of
+the topic, followed by raft elections for all partitions within the
+topic.
+
+These concepts are not visible to applications.  They are optionally visible
+to system administrators optionally:
+- system monitoring telemetry, where we would surface indicators for how many partitions
+  and topics are currently hibernated.
+- tunable configuration properties, where they could control the delay til hibernation
+  if needed.
+
+
+# Reference-level explanation
+
+## High level design
+
+The topic table, which stores the definition of the topics & partitions
+on the system, is driven by controller log commands.  New commands
+are added for hibernate and wake.  Nodes in a partition's raft group
+notice the hibernation of the topic, and shut down the raft group.
+
+There is a "hibernating" intermediate state where members of the
+partitions' raft groups are responsible for refusing writes, and
+reporting the highest offset back to the controller so that it can
+reliably store the frozen offset state of the topic when storing
+the controller message that finalizes the hibernation.
+
+Once the topic is hibernated, all its partitions will report a
+fictional leader, so that clients know where to send metadata requests,
+and so that monitoring systems do not incorrectly flag the partition
+as being in a problematic state.
+
+On startup, a hibernated partition's storage resources are not
+recovered or replayed.  Hibernated partitions are not subject to
+continuous retention policies, BUT the controller can notice if
+a partition has been hibernated for longer than its retention.ms
+limit, and wake it up for long enough to do housekeeping.
```

Comment:
```
BUT the controller can notice if
 a partition has been hibernated for longer than its retention.ms
 limit, and wake it up for long enough to do housekeeping.
```

the housekeeping that happens in the `log_manager`, IIRC, sort of happens without explicit coordination with raft and independently on each node (for better or worse). so it may not be necessary to wake the full raft group to do this until it comes time to revamp the relationship between storage, raft, and housekeeping.

Reply:
My thinking is that we'd tear down everything for a hibernated partition, including its storage-level structures that would include it in housekeeping.  We could make the storage layer have a list of hibernated logs and hydrate those logs' structures when needed to apply retention policy, but that kind of contradicts the "don't use system resources for hibernated things" high level concept.

This is one of those areas where there's a tension between the "big hammer" topic hibernation, and implementing finer-grained handling of idle partitions in each subsystem.